Jellyfin.Networking/HappyEyeballs/HttpClientExtension.cs

```diff
@@ -0,0 +1,136 @@
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+
+namespace Jellyfin.Networking.HappyEyeballs
+{
+    /// <summary>
+    /// Defines the <see cref="HttpClientExtension"/> class.
+    ///
+    /// Implementation taken from https://github.com/ppy/osu-framework/pull/4191 .
+    /// </summary>
+    public static class HttpClientExtension
+    {
+        private const int ConnectionEstablishTimeout = 2000;
+
+        /// <summary>
+        /// Interlocked doesn't support bool types.
+        /// </summary>
+        private static int _useIPv6 = 0;
+
+        /// <summary>
+        /// Gets a value indicating whether the initial IPv6 check has been performed (to determine whether v6 is available or not).
+        /// </summary>
+        private static bool _hasResolvedIPv6Availability;
+
+        /// <summary>
+        /// Gets or sets a value indicating whether IPv6 should be preferred. Value may change based on runtime failures.
+        /// </summary>
+        public static bool UseIPv6
+        {
+            get => Interlocked.CompareExchange(ref _useIPv6, 1, 1) == 1;
+            set
+            {
+                if (value)
+                {
+                    Interlocked.CompareExchange(ref _useIPv6, 1, 0);
+                }
+                else
+                {
+                    Interlocked.CompareExchange(ref _useIPv6, 0, 1);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Implements the httpclient callback method.
+        /// </summary>
+        /// <param name="context">The <see cref="SocketsHttpConnectionContext"/> instance.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> instance.</param>
+        /// <returns>The http steam.</returns>
+        public static async ValueTask<Stream> OnConnect(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
+        {
+            // Until .NET supports an implementation of Happy Eyeballs (https://tools.ietf.org/html/rfc8305#section-2),
```

Comment:
I have been thinking about this implementation today and read the old dotnet issue. This implementation doesn't really follow the RFC spec as the RFC states that both IPv6 and IPv4 should be called in parallel and whichever finishes first wins. This tries IPv6 first and then IPv4. And should IPv6 be disabled by one failure or is it more nuanced?

Reply:
That is a good point I will get that changed.